### PR TITLE
Document Workflow Update internals

### PR DESCRIPTION
**Pull request by alexshtin**
## What changed?

Document Workflow Update internals.

## Why?

For better visibility on how Workflow Update is implemented.

### CODE REVIEW

File: docs/effect-package.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
+# `effect` package
+
+## Motivation
+Acceptance or competition of Workflow Update involves 2 steps:
```
Comment (stephanos) [inline docs/effect-package.md:4]:
```suggestion
Acceptance or completion of Workflow Update involves 2 steps:
```

File: docs/effect-package.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
+# `effect` package
```
Comment (stephanos) [inline docs/effect-package.md:1]: Should this just be in the actual package itself; inside a `package.go` file? Similar to https://github.com/temporalio/temporal/blob/main/internal/goro/package.go

File: docs/effect-package.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+It is important to point out that `effect` package doesn't provide any transaction guarantees,
+and if `effect.Buffer` is lost, entire batch can be partially completed (which is acceptable
+for a Workflow Update case).  
+
+## Usage
+This package was created for Workflow Update feature, and currently used only there. It is not coupled
+with Workflow Update though, and can be used anywhere where delayed execution is required. 
+
+> #### TODO
+> `workflowTaskResponseMutation` is one of the candidates to migrate to `effect` package.
```
Comment (stephanos) [inline docs/effect-package.md:34]: I think I would leave todos out of docs. Seems misplaced to me

File: docs/effect-package.md (hunk 2 above)
Comment (alexshtin) [reply docs/effect-package.md:1]: I believe we agreed that all actual docs should be in `docs` root dir. Only `README.md` are allowed in specific dirs and they should have only link to file in `docs`. I will add `README.md` there.

File: docs/effect-package.md (hunk 3 above)
Comment (alexshtin) [reply docs/effect-package.md:34]: I used same approach everywhere. To indicate intend, or how the feature can be extended.

File: docs/effect-package.md (hunk 2 above)
Comment (stephanos) [reply docs/effect-package.md:1]: We did say that 👍 at the same time, I think of the `docs/` as high-level, conceptual documentation. This one feels more like a low-level, code comment (with some extra context).

I have no objections to leaving it here.

File: docs/effect-package.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+succeeds and `Cancel()` if it failed (usually in `deffer` block). Order of execution is the same
+as register order for `Apply()`, and reversed for `Cancel()`.
+
+It is important to point out that `effect` package doesn't provide any transaction guarantees,
+and if `effect.Buffer` is lost, entire batch can be partially completed (which is acceptable
+for a Workflow Update case).  
+
+## Usage
+This package was created for Workflow Update feature, and currently used only there. It is not coupled
+with Workflow Update though, and can be used anywhere where delayed execution is required.
```
[the remark this answers is not in the document]
Comment (stephanos) [reply docs/effect-package.md:31]: Seems like a comment that is easily going out of date ;) And something an IDE can point out, too. 🤷‍♂️

File: docs/effect-package.md (hunk 3 above)
Comment (stephanos) [reply docs/effect-package.md:34]: I can see how todos for the feature itself can be useful; but this particular todo is about another feature that can use this package. Which I think is very different.

File: docs/architecture/effect-package.md [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+> It is important to point out that `effect` package doesn't provide any transaction guarantees.
+> For example `effect.Buffer` can be partially applied (or not applied at all) after persistence
+> operation completed successfully (which is acceptable for a Workflow Update case).  
+
+## Usage
+This package was created for Workflow Update feature, and currently used only there. It is not coupled
+with Workflow Update though, and can be used anywhere where delayed execution is required. 
+
+> #### TODO
+> `workflowTaskResponseMutation` is one of the candidates to migrate to `effect` package.
```
Comment (stephanos) [inline docs/architecture/effect-package.md:35]:
```suggestion
# `effect` package

## Motivation
When processing an operation, Temporal's persistence accumulates changes to the mutable state and only
commits them to the database at the end of the operation. For an Update Workflow, the caller of the
`UpdateWorkflowExecution` API can only be unblocked after the database write has successfully completed, ensuring that the update is durable.

Unblocking the caller immediately after processing the worker's Update result in `RespondWorkflowTaskCompleted` would introduce a poor design by tightly coupling these two separate concerns.

To address this, the effect package was introduced.

## How it works
`effect.Buffer` allows to accumulate callbacks - in the same way persistence
accumulates changes - and exposes methods to `Apply()` or `Cancel()` them.

For each operation, two callbacks should be registered: one for the success case, and one for when 
the operation needs to be canceled. Then, the `Apply()` method should be called right after the
operation succeeded - or `Cancel()` when it failed (usually in the `defer` block). The order the callbacks
are executed in is the same order they were registered when using `Apply()`, and reversed for `Cancel()`.

#### NOTE
> It is important to note that the `effect` package doesn't provide any transactional guarantees.
> For example, the callbacks in `effect.Buffer` can be partially applied (or not applied at all)
> after the persistence write completed successfully (which is acceptable for a Workflow Update case).  

## Usage
This package was created specifically for the Workflow Update feature, and is currently used only there.
It is not coupled with Workflow Update, though, and can be used anywhere where delayed execution is required. 

> #### TODO
> `workflowTaskResponseMutation` is one of the candidates to migrate to `effect` package.
```

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -2,7 +2,7 @@
 
 This doc is for contributors to Temporal Server (hopefully that's you!)
 
-**Note:** All contributors also need to fill out the [Temporal Contributor License Agreement](docs/development/temporal-cla.md) before we can merge in any of your changes.
+**Note:** All contributors also need to fill out the [Temporal Contributor License Agreement](./docs/development/temporal-cla.md) before we can merge in any of your changes.
```
[the remark this answers is not in the document]
Comment (alexshtin) [reply CONTRIBUTING.md:5]: It does. There are dot in few other places, I just decided to unify it, as I was checking all of them.

**Review by stephanos — APPROVED** [review]

**Pull request merged**